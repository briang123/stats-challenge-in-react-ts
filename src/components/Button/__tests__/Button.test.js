import React from 'react';
import { cleanup, render } from '@testing-library/react';
import { create } from 'react-test-renderer';
import Button from '../Button';

afterEach(() => {
  cleanup();
  console.error.mockClear();
});

console.error = jest.fn();
const mockFn = jest.fn();

describe('<Button />', () => {
  it('Button component without onClick prop should log error and not render button', () => {
    const { queryByTitle } = render(<Button />);
    expect(console.error).toHaveBeenCalled();
    const button = queryByTitle('BUTTON');
    expect(button).toBeNull();
  });

  it('Button component with an invalid onClick function prop should log error and not render button', () => {
    const { queryByTitle } = render(<Button onClick="" />);
    expect(console.error).toHaveBeenCalled();
    const button = queryByTitle('BUTTON');
    expect(button).toBeNull();
  });

  it('Button component with an valid onClick function and no other props should render button', () => {
    const { container } = render(<Button onClick={mockFn} />);
    expect(console.error).not.toHaveBeenCalled();
    const btn = container.querySelector('button');
    expect(btn).not.toBeNull();
  });

  it('Button component with an valid onClick and text value for props should show button text', () => {
    const { container } = render(<Button onClick={mockFn} text="foo" />);
    const btn = container.querySelector('button');
    expect(btn.textContent).toBe('foo');
  });

  it('Button component with an valid onClick and text value for props should show button text', () => {
    const { getByTestId } = render(<Button text="foo" testid="bar" onClick={mockFn} />);
    const btn = getByTestId('bar');
    expect(btn.textContent).toBe('foo');
  });

  it('Button component should be enabled by default', () => {
    const { getByTestId } = render(<Button text="foo" testid="bar" onClick={mockFn} />);
    const btn = getByTestId('bar');
    expect(btn.disabled).toBe(false);
  });

  it('Button component should have tabIndex=-1 by default', () => {
    const { getByTestId } = render(<Button text="foo" testid="bar" onClick={mockFn} />);
    const btn = getByTestId('bar');
    expect(btn.tabIndex).toBe(-1);
  });

  it('Button component should be of btnType default by default', () => {
    const { getByTestId } = render(<Button text="foo" testid="bar" onClick={mockFn} />);
    const btn = getByTestId('bar');
    expect(window.getComputedStyle(btn)._values['border-radius']).toBe('44px');
  });

  it('Button component should reflect the btnType when set to submit', () => {
    const { getByTestId } = render(
      <Button text="foo" testid="bar" btnType="submit" onClick={mockFn} />
    );
    const btn = getByTestId('bar');
    expect(window.getComputedStyle(btn)._values['border-radius']).toBe('0px 20px 20px 0px');
  });

  it('Button component should be disabled when we set it to be', () => {
    const { getByTestId } = render(
      <Button text="foo" testid="bar" disabled={true} onClick={mockFn} />
    );
    const btn = getByTestId('bar');
    expect(btn.disabled).toBe(true);
  });

  it('Should match snapshot with props', () => {
    const tree = create(
      <Button text="foo" testid="bar" disabled={true} onClick={mockFn} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
