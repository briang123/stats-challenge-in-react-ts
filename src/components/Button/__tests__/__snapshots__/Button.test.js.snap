// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Should match snapshot with props 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  border: none;
  height: 4.87rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: var(--lightcolor);
  outline: none;
  border-radius: 44px;
  width: 375px;
  background-color: var(--defaultbtn);
}

.c0:hover {
  cursor: pointer;
}

<button
  className="c0"
  data-testid="bar"
  disabled={true}
  onClick={[MockFunction]}
  tabIndex={-1}
>
  foo
</button>
`;
